File: internal/domain/template/spec.go
package template

import (
	"encoding/json"
	"time"

	"github.com/davidmovas/postulator/internal/domain/llm"
)

type Scope string

const (
	ScopeGlobal Scope = "global"
	ScopeSite   Scope = "site"
)

func (s Scope) Valid() bool {
	switch s {
	case ScopeGlobal, ScopeSite:
		return true
	default:
		return false
	}
}

type ImageSource string

const (
	ImagesAI      ImageSource = "ai"
	ImagesWPMedia ImageSource = "wpmedia"
	ImagesLocal   ImageSource = "local"
)

func (s ImageSource) Valid() bool {
	switch s {
	case ImagesAI, ImagesWPMedia, ImagesLocal:
		return true
	default:
		return false
	}
}

type AnchorStrategy string

const (
	AnchorPreferUser AnchorStrategy = "prefer_user"
	AnchorRotate     AnchorStrategy = "rotate"
)

func (s AnchorStrategy) Valid() bool {
	switch s {
	case AnchorPreferUser, AnchorRotate:
		return true
	default:
		return false
	}
}

type OverrideScope string

const (
	OverrideSite OverrideScope = "site"
	OverridePage OverrideScope = "page"
)

func (s OverrideScope) Valid() bool {
	switch s {
	case OverrideSite, OverridePage:
		return true
	default:
		return false
	}
}

type SectionKeywordRules struct {
	Include          []string `json:"include" description:"Phrases this section must use at least once"`
	PrimaryInHeading bool     `json:"primaryInHeading" description:"The primary keyword must appear in this section's heading"`
}

type Section struct {
	Heading      string              `json:"heading" description:"The heading; {primaryKeyword}, {entityName}, {siteName} and {pageTitle} are filled in per page"`
	Intent       string              `json:"intent" description:"What the section has to cover, one short sentence to the writer"`
	TargetWords  int                 `json:"targetWords" minimum:"0" description:"About how many words the section should run to"`
	Required     bool                `json:"required" description:"The page is not valid without this section"`
	KeywordRules SectionKeywordRules `json:"keywordRules" description:"What this section has to say about the keywords"`
}

type Length struct {
	Min int `json:"min" minimum:"0" description:"The fewest words the whole page may run to"`
	Max int `json:"max" minimum:"0" description:"The most words the whole page may run to"`
}

type KeywordRules struct {
	PrimaryInTitle          bool    `json:"primaryInTitle" description:"The primary keyword must appear in the title"`
	PrimaryInH1             bool    `json:"primaryInH1" description:"The primary keyword must appear in the first heading"`
	PrimaryInFirstParagraph bool    `json:"primaryInFirstParagraph" description:"The primary keyword must appear in the opening paragraph"`
	MaxDensity              float64 `json:"maxDensity" minimum:"0" maximum:"1" description:"The largest share of the words the primary keyword may take, between 0 and 1"`
}

type LinkRules struct {
	UpDepth                    int     `json:"upDepth" minimum:"0" description:"How many levels up the tree a page links to, 1 for its parent alone"`
	DownLinks                  bool    `json:"downLinks" description:"Link down to the children of the entity"`
	SiblingMinWeight           float64 `json:"siblingMinWeight" minimum:"0" maximum:"1" description:"The weight a related edge needs before a sibling link is placed, between 0 and 1"`
	MaxLinks                   int     `json:"maxLinks" minimum:"0" description:"The most internal links one page may carry"`
	MaxPerTarget               int     `json:"maxPerTarget" minimum:"0" description:"The most links one page may point at a single target"`
	ParentLinkWithinParagraphs int     `json:"parentLinkWithinParagraphs" minimum:"0" description:"The parent link must appear within this many paragraphs of the start"`
	ChildrenSection            bool    `json:"childrenSection" description:"Close the page with a section listing its children"`
}

type MetaRules struct {
	TitlePattern   string `json:"titlePattern" description:"How to build the SEO title, for example {primaryKeyword} | {siteName}; {entityName} and {pageTitle} work too"`
	DescriptionMax int    `json:"descriptionMax" minimum:"0" description:"The most characters the SEO description may run to"`
}

type Images struct {
	Featured bool        `json:"featured" description:"The page carries a featured image"`
	Inline   int         `json:"inline" minimum:"0" description:"How many images to place inside the body; zero for a page without images"`
	Source   ImageSource `json:"source" enum:"ai,wpmedia,local" description:"Where the images come from: drawn by a model, picked from the WordPress library, or read from a folder"`
}

func (i Images) Wanted() int {
	count := max(i.Inline, 0)
	if i.Featured {
		count++
	}
	return count
}

type StepSpec struct {
	Name    string         `json:"name"`
	Enabled bool           `json:"enabled"`
	Params  map[string]any `json:"params,omitempty"`
}

type TemplateSpec struct {
	Sections      []Section                 `json:"sections"`
	Tone          string                    `json:"tone"`
	Length        Length                    `json:"length"`
	KeywordRules  KeywordRules              `json:"keywordRules"`
	LinkRules     LinkRules                 `json:"linkRules"`
	MetaRules     MetaRules                 `json:"metaRules"`
	Images        Images                    `json:"images"`
	ModelProfiles map[llm.Role]llm.ModelRef `json:"modelProfiles"`
	Recipe        []StepSpec                `json:"recipe"`
}

type Template struct {
	ID        string
	Scope     Scope
	SiteID    *string
	Name      string
	PageKind  string
	Version   int
	Spec      TemplateSpec
	CreatedAt time.Time
	UpdatedAt time.Time
}

type Override struct {
	ID         string
	TemplateID string
	Scope      OverrideScope
	TargetID   string
	Patch      json.RawMessage
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

type LinkPolicy struct {
	ID             string
	Scope          Scope
	SiteID         *string
	Name           string
	Rules          LinkRules
	ForbidExternal bool
	ForbidSelf     bool
	AnchorStrategy AnchorStrategy
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

File: frontend/src/features/templates/spec.ts
import type { TemplateSpec } from "../../data/types.js";
import type { JsonObject, JsonValue } from "../../domain/merge-patch.js";
import { modelRoles, stepNames } from "../../generated/vocab.js";

export interface SectionDraft {
    heading: string;
    intent: string;
    targetWords: number;
    required: boolean;
    include: string[];
    primaryInHeading: boolean;
}

export interface ProfileDraft {
    role: string;
    provider: string;
    model: string;
}

export interface StepDraft {
    name: string;
    enabled: boolean;
    declared: boolean;
    params: JsonObject | null;
}

export interface SpecDraft {
    sections: SectionDraft[];
    tone: string;
    lengthMin: number;
    lengthMax: number;
    primaryInTitle: boolean;
    primaryInH1: boolean;
    primaryInFirstParagraph: boolean;
    maxDensity: number;
    upDepth: number;
    downLinks: boolean;
    siblingMinWeight: number;
    maxLinks: number;
    maxPerTarget: number;
    parentLinkWithinParagraphs: number;
    childrenSection: boolean;
    titlePattern: string;
    descriptionMax: number;
    featuredImage: boolean;
    inlineImages: number;
    imageSource: string;
    profiles: ProfileDraft[];
    recipe: StepDraft[];
}

export const imageStep = "generate_images";
const drawnSource = "ai";

export function isJsonObject(value: JsonValue | undefined): value is JsonObject {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

function branch(parent: JsonObject | null, key: string): JsonObject | null {
    if (parent === null) {
        return null;
    }
    const held = parent[key];
    return isJsonObject(held) ? held : null;
}

function text(parent: JsonObject | null, key: string): string {
    const held = parent === null ? undefined : parent[key];
    return typeof held === "string" ? held : "";
}

function count(parent: JsonObject | null, key: string): number {
    const held = parent === null ? undefined : parent[key];
    return typeof held === "number" && Number.isFinite(held) ? held : 0;
}

function flag(parent: JsonObject | null, key: string): boolean {
    const held = parent === null ? undefined : parent[key];
    return held === true;
}

function series(parent: JsonObject | null, key: string): JsonValue[] {
    const held = parent === null ? undefined : parent[key];
    return Array.isArray(held) ? held : [];
}

function words(parent: JsonObject | null, key: string): string[] {
    return series(parent, key).filter((held): held is string => typeof held === "string");
}

export function jsonOf(spec: TemplateSpec): JsonValue {
    const encoded: JsonValue = JSON.parse(JSON.stringify(spec));
    return encoded;
}

function sectionOf(value: JsonValue): SectionDraft {
    const held = isJsonObject(value) ? value : null;
    const rules = branch(held, "keywordRules");
    return {
        heading: text(held, "heading"),
        intent: text(held, "intent"),
        targetWords: count(held, "targetWords"),
        required: flag(held, "required"),
        include: words(rules, "include"),
        primaryInHeading: flag(rules, "primaryInHeading"),
    };
}

function profilesOf(held: JsonObject | null): ProfileDraft[] {
    if (held === null) {
        return [];
    }
    const known = modelRoles.filter((role) => Object.prototype.hasOwnProperty.call(held, role));
    const extra = Object.keys(held).filter((role) => !(modelRoles as readonly string[]).includes(role));
    return [...known, ...extra].map((role) => {
        const ref = branch(held, role);
        return { role, provider: text(ref, "provider"), model: text(ref, "model") };
    });
}

function recipeOf(rows: JsonValue[]): StepDraft[] {
    const params = new Map<string, JsonObject | null>();
    const enabled = new Map<string, boolean>();
    const order: string[] = [];
    for (const row of rows) {
        if (!isJsonObject(row)) {
            continue;
        }
        const name = text(row, "name");
        if (name === "" || order.includes(name)) {
            continue;
        }
        order.push(name);
        params.set(name, branch(row, "params"));
        enabled.set(name, flag(row, "enabled"));
    }
    const extra = order.filter((name) => !(stepNames as readonly string[]).includes(name));
    return [...stepNames, ...extra].map((name) => ({
        name,
        enabled: enabled.get(name) === true,
        declared: order.includes(name),
        params: params.get(name) ?? null,
    }));
}

export function draftFromJson(value: JsonValue): SpecDraft {
    const root = isJsonObject(value) ? value : null;
    const length = branch(root, "length");
    const keywords = branch(root, "keywordRules");
    const links = branch(root, "linkRules");
    const meta = branch(root, "metaRules");
    const images = branch(root, "images");
    return {
        sections: series(root, "sections").map(sectionOf),
        tone: text(root, "tone"),
        lengthMin: count(length, "min"),
        lengthMax: count(length, "max"),
        primaryInTitle: flag(keywords, "primaryInTitle"),
        primaryInH1: flag(keywords, "primaryInH1"),
        primaryInFirstParagraph: flag(keywords, "primaryInFirstParagraph"),
        maxDensity: count(keywords, "maxDensity"),
        upDepth: count(links, "upDepth"),
        downLinks: flag(links, "downLinks"),
        siblingMinWeight: count(links, "siblingMinWeight"),
        maxLinks: count(links, "maxLinks"),
        maxPerTarget: count(links, "maxPerTarget"),
        parentLinkWithinParagraphs: count(links, "parentLinkWithinParagraphs"),
        childrenSection: flag(links, "childrenSection"),
        titlePattern: text(meta, "titlePattern"),
        descriptionMax: count(meta, "descriptionMax"),
        featuredImage: flag(images, "featured"),
        inlineImages: count(images, "inline"),
        imageSource: text(images, "source"),
        profiles: profilesOf(branch(root, "modelProfiles")),
        recipe: recipeOf(series(root, "recipe")),
    };
}

export function draftOf(spec: TemplateSpec): SpecDraft {
    return draftFromJson(jsonOf(spec));
}

export function specJsonOf(draft: SpecDraft): JsonObject {
    const profiles: JsonObject = {};
    for (const profile of draft.profiles) {
        profiles[profile.role] = { provider: profile.provider, model: profile.model };
    }
    const recipe: JsonValue[] = [];
    for (const step of draft.recipe) {
        if (!step.enabled && !step.declared) {
            continue;
        }
        const row: JsonObject = { name: step.name, enabled: step.enabled };
        if (step.params !== null && Object.keys(step.params).length > 0) {
            row["params"] = step.params;
        }
        recipe.push(row);
    }
    return {
        sections: draft.sections.map((section) => ({
            heading: section.heading,
            intent: section.intent,
            targetWords: section.targetWords,
            required: section.required,
            keywordRules: { include: [...section.include], primaryInHeading: section.primaryInHeading },
        })),
        tone: draft.tone,
        length: { min: draft.lengthMin, max: draft.lengthMax },
        keywordRules: {
            primaryInTitle: draft.primaryInTitle,
            primaryInH1: draft.primaryInH1,
            primaryInFirstParagraph: draft.primaryInFirstParagraph,
            maxDensity: draft.maxDensity,
        },
        linkRules: {
            upDepth: draft.upDepth,
            downLinks: draft.downLinks,
            siblingMinWeight: draft.siblingMinWeight,
            maxLinks: draft.maxLinks,
            maxPerTarget: draft.maxPerTarget,
            parentLinkWithinParagraphs: draft.parentLinkWithinParagraphs,
            childrenSection: draft.childrenSection,
        },
        metaRules: { titlePattern: draft.titlePattern, descriptionMax: draft.descriptionMax },
        images: { featured: draft.featuredImage, inline: draft.inlineImages, source: draft.imageSource },
        modelProfiles: profiles,
        recipe,
    };
}

export function specOf(draft: SpecDraft): TemplateSpec {
    return specJsonOf(draft) as unknown as TemplateSpec;
}

export function asksForImages(draft: SpecDraft): boolean {
    return draft.featuredImage || draft.inlineImages > 0;
}

export function readyForImages(before: SpecDraft, next: SpecDraft): SpecDraft {
    if (asksForImages(before) || !asksForImages(next)) {
        return next;
    }
    const ownRecipe = next.recipe.some((step) => step.enabled || step.declared);
    return {
        ...next,
        imageSource: next.imageSource === "" ? drawnSource : next.imageSource,
        recipe: ownRecipe
            ? next.recipe.map((step) => (step.name === imageStep ? { ...step, enabled: true, declared: true } : step))
            : next.recipe,
    };
}

export function emptySection(): SectionDraft {
    return { heading: "", intent: "", targetWords: 0, required: true, include: [], primaryInHeading: false };
}

export function moved<T>(items: readonly T[], from: number, to: number): T[] {
    if (from < 0 || to < 0 || from >= items.length || to >= items.length) {
        return [...items];
    }
    const out = [...items];
    const held = out[from];
    if (held === undefined) {
        return [...items];
    }
    out.splice(from, 1);
    out.splice(to, 0, held);
    return out;
}
